compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -121,15 +122,22 @@ impl<Tag> Allocation<Tag> {
         Allocation::from_bytes(slice, Align::ONE, Mutability::Not)
     }
 
-    pub fn uninit(size: Size, align: Align) -> Self {
-        Allocation {
-            bytes: vec![0; size.bytes_usize()],
+    /// Try to create an Allocation of `size` bytes, failing if there is not enough memory
+    /// available to the compiler to do so.
+    pub fn uninit(size: Size, align: Align) -> InterpResult<'static, Self> {
+        let mut bytes = Vec::new();
+        bytes.try_reserve(size.bytes_usize()).map_err(|_| {
+            InterpError::ResourceExhaustion(ResourceExhaustionInfo::MemoryExhausted)
```

Comment:
Please add a comment here that this is subtle because we are letting external factors (the system memory allocator) affect the result of a query. The only reason this is okay is that the only possible effect here is a hard error that will abort compilation.